### PR TITLE
[CPU] ScatterUpdate ScatterElementsUpdate and ScatterNDUpdate support

### PR DESCRIPTION
**Pull request by chenhu-wang**
[CPU] add support in mkldnn plugin
[TEST] single layer test

### CODE REVIEW

File: inference-engine/src/mkldnn_plugin/nodes/mkldnn_scatter_update_node.cpp [hunk 1]
```diff
@@ -0,0 +1,531 @@
...
+MKLDNNScatterUpdateNode::MKLDNNScatterUpdateNode(const InferenceEngine::CNNLayerPtr& layer, const mkldnn::engine& eng, MKLDNNWeightsSharing::Ptr &cache)
+        : MKLDNNNode(layer, eng, cache), inputSize(0lu), indicesSize(0lu), outputSize(0lu), axisSize(0lu),
+        inputPrec(Precision::UNSPECIFIED), indicesPrec(Precision::UNSPECIFIED), outputPrec(Precision::UNSPECIFIED), axisPrec(Precision::UNSPECIFIED) {}
+
+void MKLDNNScatterUpdateNode::getSupportedDescriptors() {
+    if (!descs.empty())
+        return;
+
+    if ((getParentEdges().size() != 3) && (getParentEdges().size() != 4))
+        THROW_IE_EXCEPTION << "Incorrect number of input edges for layer " << getName();
```
[the remark this answers is not in the document]
Comment (chenhu-wang) [reply inference-engine/src/mkldnn_plugin/nodes/mkldnn_scatter_update_node.cpp]: fixed for all exception message.

File: inference-engine/src/mkldnn_plugin/nodes/mkldnn_scatter_update_node.cpp [hunk 2]
```diff
@@ -0,0 +1,531 @@
...
+    if (getChildEdges().empty())
+        THROW_IE_EXCEPTION << "Incorrect number of output edges for layer " << getName();
+
+    if (getParentEdgeAt(0)->getDims().ndims() < 1 || getParentEdgeAt(0)->getDims().ndims() > 5 ||
+        getParentEdgeAt(1)->getDims().ndims() < 1 || getParentEdgeAt(1)->getDims().ndims() > 5 ||
+        getParentEdgeAt(2)->getDims().ndims() < 1 || getParentEdgeAt(2)->getDims().ndims() > 5) {
+        THROW_IE_EXCEPTION << "Only tensor of rank 1 to 5 is supported for layer " << getName();
+    }
```
[the remark this answers is not in the document]
Comment (chenhu-wang) [reply inference-engine/src/mkldnn_plugin/nodes/mkldnn_scatter_update_node.cpp]: added common check and specific check, and put these check in init stage as much as possible.

File: inference-engine/src/mkldnn_plugin/nodes/mkldnn_scatter_update_node.cpp [hunk 3]
```diff
@@ -0,0 +1,531 @@
...
+
+    auto indicesPre = getCnnLayer()->insData[INDICES_ID].lock()->getPrecision();
+    if (indicesPre != Precision::I32 && indicesPre != Precision::I64) {
```
[the remark this answers is not in the document]
Comment (chenhu-wang) [reply inference-engine/src/mkldnn_plugin/nodes/mkldnn_scatter_update_node.cpp]: fixed.

File: inference-engine/src/mkldnn_plugin/nodes/mkldnn_scatter_update_node.cpp [hunk 4]
```diff
@@ -0,0 +1,531 @@
...
+    Precision inputPrecision = getCnnLayer()->insData[DATA_ID].lock()->getPrecision();
+    Precision updatePrecision = getCnnLayer()->insData[UPDATE_ID].lock()->getPrecision();
+    Precision outputPrecision = getCnnLayer()->outData[0]->getPrecision();
+    if (inputPrecision != updatePrecision) {
+        THROW_IE_EXCEPTION << "Input precision and update precision are not the same for layer " << getName();
+    }
+    if (inputPrecision != outputPrecision) {
+        outputPrecision = inputPrecision;
+    }
+    if (!isOneOf(inputPrecision, {Precision::FP32, Precision::FP16, Precision::BF16, Precision::I64, Precision::U64, Precision::I32,
```
[the remark this answers is not in the document]
Comment (chenhu-wang) [reply inference-engine/src/mkldnn_plugin/nodes/mkldnn_scatter_update_node.cpp]: fixed, now avoid using explicit data type.

File: inference-engine/src/mkldnn_plugin/nodes/mkldnn_scatter_update_node.cpp [hunk 5]
```diff
@@ -0,0 +1,531 @@
...
+        for (auto p : precisions) {
+            if (p == precision) {
+                return true;
+            }
+        }
+        return false;
+    };
+
+    Precision indicesPrecision = getCnnLayer()->insData[INDICES_ID].lock()->getPrecision();
+    if (!isOneOf(indicesPrecision, {Precision::I32, Precision::I64})) {
```
[the remark this answers is not in the document]
Comment (chenhu-wang) [reply inference-engine/src/mkldnn_plugin/nodes/mkldnn_scatter_update_node.cpp]: Fixed for indices and axis.
And also relaxed update data type restriction(should same as input) by same machanism. If update data type is not same as input, convert as well. Input, output and update configued as same data type in init primitive description.

File: inference-engine/src/mkldnn_plugin/nodes/mkldnn_scatter_update_node.cpp [hunk 6]
```diff
@@ -0,0 +1,531 @@
...
+}
+
+void MKLDNNScatterUpdateNode::createPrimitive() {
+    auto &dstMemPtr = getChildEdgeAt(0)->getMemoryPtr();
+    auto &srcMemPtr = getParentEdgeAt(DATA_ID)->getMemoryPtr();
+    auto &indicesMemPtr = getParentEdgeAt(INDICES_ID)->getMemoryPtr();
+    auto &updateMemPtr = getParentEdgeAt(UPDATE_ID)->getMemoryPtr();
+
+    if (!dstMemPtr || !dstMemPtr->GetPrimitivePtr())
+        THROW_IE_EXCEPTION << "Destination memory didn't allocate.";
```
[the remark this answers is not in the document]
Comment (chenhu-wang) [reply inference-engine/src/mkldnn_plugin/nodes/mkldnn_scatter_update_node.cpp]: fixed

File: inference-engine/src/mkldnn_plugin/nodes/mkldnn_scatter_update_node.cpp [hunk 7]
```diff
@@ -0,0 +1,531 @@
...
+        if (axisSize == 4) {
+            auto *axisPtr32 = reinterpret_cast<int32_t*>(axisPtr);
+            axis = *axisPtr32;
+        } else {
+            auto *axisPtr64 = reinterpret_cast<int64_t*>(axisPtr);
+            axis = *axisPtr64;
+        }
+    }
+
+    if (inputSize == 1) {
```
[the remark this answers is not in the document]
Comment (chenhu-wang) [reply inference-engine/src/mkldnn_plugin/nodes/mkldnn_scatter_update_node.cpp]: fixed

File: inference-engine/src/mkldnn_plugin/nodes/mkldnn_scatter_update_node.cpp [hunk 8]
```diff
@@ -0,0 +1,531 @@
...
+    }
+    return block5D;
+}
+
+template <typename data_t, typename index_t>
+void MKLDNNScatterUpdateNode::scatterUpdate(data_t *srcData, index_t *indices, data_t *update, int axis, data_t *dstData, ScatterUpdateMode mode) {
+    SizeVector srcDataDim = getParentEdgeAt(0)->getDesc().getDims();
+    SizeVector indicesDim = getParentEdgeAt(1)->getDesc().getDims();
+    SizeVector updateDim = getParentEdgeAt(2)->getDesc().getDims();
+    SizeVector dstDataDim = getChildEdgeAt(0)->getDesc().getDims();
```
[the remark this answers is not in the document]
Comment (chenhu-wang) [reply inference-engine/src/mkldnn_plugin/nodes/mkldnn_scatter_update_node.cpp]: fixed

File: inference-engine/src/mkldnn_plugin/nodes/mkldnn_scatter_update_node.cpp [hunk 9]
```diff
@@ -0,0 +1,531 @@
...
+
+    size_t srcRank = srcDataDim.size();
+    size_t dstRank = dstDataDim.size();
+    size_t indicesRank = indicesDim.size();
+    size_t updateRank = updateDim.size();
+
+    // common check
+    if (srcRank != dstRank) {
```
[the remark this answers is not in the document]
Comment (chenhu-wang) [reply inference-engine/src/mkldnn_plugin/nodes/mkldnn_scatter_update_node.cpp]: fixed!

File: inference-engine/src/mkldnn_plugin/nodes/mkldnn_scatter_update_node.cpp [hunk 10]
```diff
@@ -0,0 +1,531 @@
...
+
+        SizeVector expectUpdateShape = {};
+        size_t tupleRank = indicesRank - 1;
+        for (size_t ri = 0; ri < tupleRank; ri++) {
+            expectUpdateShape.push_back(indicesDim[ri]);
+        }
+        for (size_t rd = k; rd < srcRank; rd++) {
+            expectUpdateShape.push_back(srcDataDim[rd]);
+        }
+        if (expectUpdateShape.size() != updateRank) {
```
[the remark this answers is not in the document]
Comment (chenhu-wang) [reply inference-engine/src/mkldnn_plugin/nodes/mkldnn_scatter_update_node.cpp]: Moved [most checks](https://github.com/openvinotoolkit/openvino/pull/909/files#diff-055684644d4488c8b03ebe7511a8c0b7R87) to initSupportedPrimitiveDescriptors() .
[Some check](https://github.com/openvinotoolkit/openvino/pull/909/files#diff-055684644d4488c8b03ebe7511a8c0b7R336) need exact value of axis tensor.  I leave to execute(). I am not sure we can get tensor value in init stage. Please correct me I am wrong.

File: inference-engine/src/mkldnn_plugin/nodes/mkldnn_scatter_update_node.cpp [hunk 11]
```diff
@@ -0,0 +1,531 @@
...
+void MKLDNNScatterUpdateNode::getSupportedDescriptors() {
+    if (!descs.empty())
+        return;
+
+    if ((getParentEdges().size() != 3) && (getParentEdges().size() != 4))
+        THROW_IE_EXCEPTION << "Incorrect number of input edges for layer " << getName();
+    if (getChildEdges().empty())
+        THROW_IE_EXCEPTION << "Incorrect number of output edges for layer " << getName();
+
+    if (getParentEdgeAt(0)->getDims().ndims() < 1 || getParentEdgeAt(0)->getDims().ndims() > 5 ||
```
[the remark this answers is not in the document]
Comment (chenhu-wang) [reply inference-engine/src/mkldnn_plugin/nodes/mkldnn_scatter_update_node.cpp]: extended to any dimension! thanks.

File: inference-engine/src/mkldnn_plugin/nodes/mkldnn_scatter_update_node.cpp [hunk 12]
```diff
@@ -0,0 +1,531 @@
...
+    for (size_t ri = 0; ri < indicesRank; ri++) {
+        idxLength *= indicesDim[ri];
+    }
+
+    // if src and dst is the same pointer, no need copy value
+    if (srcData != dstData) {
+        std::memcpy(dstData, srcData, shape5D[0] * block5D[0] * sizeof(data_t));
+    }
+
+    if (mode == ScatterUpdateMode::ScatterUpdate) {
```
[the remark this answers is not in the document]
Comment (chenhu-wang) [reply inference-engine/src/mkldnn_plugin/nodes/mkldnn_scatter_update_node.cpp]: fixed. thanks.

File: inference-engine/src/mkldnn_plugin/nodes/mkldnn_scatter_update_node.cpp [hunk 13]
```diff
@@ -0,0 +1,531 @@
...
+        }
+        for (size_t ru = 0; ru < updateRank; ru++) {
+            if (updateDim[ru] != expectUpdateShape[ru]) {
+                THROW_IE_EXCEPTION << "Shape of update tensor is not matched for input and indices for layer " << getName()
+                                   << "with type" << getType();
+            }
+        }
+
+        switch (axis) {
+            case 0:
```
[the remark this answers is not in the document]
Comment (chenhu-wang) [reply inference-engine/src/mkldnn_plugin/nodes/mkldnn_scatter_update_node.cpp]: Great comments! thank you. United all cases.

File: inference-engine/src/mkldnn_plugin/nodes/mkldnn_scatter_update_node.cpp [hunk 14]
```diff
@@ -0,0 +1,531 @@
...
+                parallel_for(idxLength, [&](size_t idx) {
+                    data_t *dstEntry = dstData + indices[idx] * block5D[axis];
```
[the remark this answers is not in the document]
Comment (chenhu-wang) [reply inference-engine/src/mkldnn_plugin/nodes/mkldnn_scatter_update_node.cpp]: Added [these checks](https://github.com/openvinotoolkit/openvino/pull/909/files#diff-055684644d4488c8b03ebe7511a8c0b7R309) by parallel going through and check indices value

File: inference-engine/src/mkldnn_plugin/nodes/mkldnn_scatter_update_node.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (chenhu-wang) [reply inference-engine/src/mkldnn_plugin/nodes/mkldnn_scatter_update_node.cpp]: thank you.  I also avoid template functions. Given this ops is just data movement.  With byte base address, we can get address offset by data size. How many byte to copy can be also get by data size.

File: ngraph/src/ngraph/op/scatter_elements_update.cpp [hunk 15]
```diff
@@ -249,10 +249,18 @@ namespace
 
         switch (out->get_element_type())
         {
+            TYPE_CASE(i8)(arg0, arg1, arg2, arg3, out, normalized_axis);
```
Comment (GlebKazantaev) [inline ngraph/src/ngraph/op/scatter_elements_update.cpp:252]: @ilyachur don't you think that this types are needed?

Comment (ilyachur) [reply ngraph/src/ngraph/op/scatter_elements_update.cpp:252]: @chenhu-wang Why do you add these types? 

We didn't face with these data types for this operation.

File: inference-engine/tests/ngraph_functions/src/scatter_ND_update.cpp [hunk 16]
```diff
@@ -0,0 +1,23 @@
...
+
+namespace ngraph {
+namespace builder {
+
+std::shared_ptr<ngraph::Node> makeScatterNDUpdate(const ngraph::Output<Node> &in,
+                                                  const element::Type& indicesType,
+                                                  const std::vector<size_t>& indicesShape,
+                                                  const std::vector<size_t>& indices,
+                                                  const ngraph::Output<Node> &update) {
+    auto indicesNode = std::make_shared<ngraph::opset1::Constant>(indicesType, indicesShape, indices);
```
Comment (ilyachur) [inline inference-engine/tests/ngraph_functions/src/scatter_ND_update.cpp:15]: Why did you add this method? It doesn't work.

File: docs/ops/movement/ScatterUpdate_3.md [hunk 17]
```diff
@@ -9,7 +9,7 @@
 **Detailed description**: *ScatterUpdate* creates a copy of the first input tensor with updated elements in positions specified with `indices` input
 and values specified with `updates` tensor starting from the dimension with index `axis`. For the `data` tensor of shape `[d_0, d_1, ..., d_n]`,
 `indices` tensor of shape `[i_0, i_1, ..., i_k]` and `updates` tensor of shape
-`[d_0, d_1, ... d_(axis - 1), i_0, i_1, ..., i_k, d_(axis + k + 1), ..., d_n]` the operation computes
+`[d_0, d_1, ... d_(axis - 1), i_0, i_1, ..., i_k, d_(axis + 1), ..., d_n]` the operation computes
```
Comment (dmitry-gorokhov) [inline docs/ops/movement/ScatterUpdate_3.md:12]: @lazarevevgeny Do you agree with this changes?

File: inference-engine/src/mkldnn_plugin/nodes/mkldnn_scatter_update_node.cpp [hunk 18]
```diff
@@ -0,0 +1,498 @@
...
+            for (size_t ru = 0; ru < updateRank; ru++) {
+                if (updateDim[ru] != expectUpdateShape[ru]) {
+                    THROW_IE_EXCEPTION << "'" << getType() << "'" << " layer with name '" << getName()
+                    << "' do not have matched tensor shape relationship for input, indices and update";
+                }
+            }
+        }
+    }
+
+    if (srcPtr != dstPtr) {
```
Comment (dmitry-gorokhov) [inline inference-engine/src/mkldnn_plugin/nodes/mkldnn_scatter_update_node.cpp:345]: In fact right this condition will always be true, so I am wondering does it make sense to enable "inplace" feature for this layer? So smt like:
`
config.inConfs[DATA_ID].inPlace = 0;
config.outConfs[0].inPlace = 0;
`
This will allow src and dst pointers to point on the same physical memory and could provide some performance benefit I suppose.

File: inference-engine/tests/ngraph_functions/src/scatter_ND_update.cpp (hunk 16 above)
Comment (chenhu-wang) [reply inference-engine/tests/ngraph_functions/src/scatter_ND_update.cpp:15]: Thank you for pointing out. I am implementing ScatterUpdate layers together. When I run single layer test, I showed ScatterNDUpdate and ScatterElementsUpdate ref are not support yet in ngraph. so I [commented out the test](https://github.com/openvinotoolkit/openvino/blob/e891d94ba09adb93bc8dbfadd59f51db8d6047ea/inference-engine/tests/functional/plugin/cpu/shared_tests_instances/single_layer_tests/scatter_ND_update.cpp#L135). and will open when support is available in ngraph.

File: ngraph/src/ngraph/op/scatter_elements_update.cpp (hunk 15 above)
Comment (chenhu-wang) [reply ngraph/src/ngraph/op/scatter_elements_update.cpp:252]: I am implementing according [ops spec](https://github.com/openvinotoolkit/openvino/blob/master/docs/ops/movement/ScatterElementsUpdate_3.md). It says input is data type of T, and T is any numeric type. So I added this support and test. Do you think It would be possible data type need support in the future? for example, If some onnx model contain [scatterElementsUpdate layer](https://github.com/onnx/onnx/blob/rel-1.7.0/docs/Operators.md#type-constraints-120), which have i8/i16/u8/u16 as input data type, we can support it directly.

File: inference-engine/tests/functional/plugin/cpu/shared_tests_instances/single_layer_tests/scatter_ND_update.cpp [hunk 19]
```diff
@@ -0,0 +1,142 @@
...
+const std::vector<std::vector<size_t>> idxValue_Depth5 = {
+        {2, 4, 5, 5, 1, 7, 9, 7, 8, 5}
+};
+const std::vector<std::vector<size_t>> updateShapes_Depth5 = {
+        {2}
+};
+
+
+const auto ScatterNDUpdate_SLICE1 = ::testing::Combine(
+        ::testing::ValuesIn(inputShapes),
```
Comment (iefode) [inline inference-engine/tests/functional/plugin/cpu/shared_tests_instances/single_layer_tests/scatter_ND_update.cpp:86]: IMO, this is such a complex design. You can make sure, that the similar case is in `squeeze_unsqueeze` single layer tests. I think, we can do the same here.

File: inference-engine/tests/functional/plugin/cpu/shared_tests_instances/single_layer_tests/scatter_ND_update.cpp [hunk 20]
```diff
@@ -0,0 +1,142 @@
...
+        ::testing::ValuesIn(idxShapes_Depth5),
+        ::testing::ValuesIn(idxValue_Depth5),
+        ::testing::ValuesIn(updateShapes_Depth5),
+        ::testing::ValuesIn(inputPrecisions),
+        ::testing::ValuesIn(idxPrecisions),
+        ::testing::Values(CommonTestUtils::DEVICE_CPU)
+);
+
+// open after ops support in ngraph merged
+// INSTANTIATE_TEST_CASE_P(ScatterNDUpdate_SLICE1, ScatterNDUpdateLayerTest, ScatterNDUpdate_SLICE1, ScatterNDUpdateLayerTest::getTestCaseName);
```
Comment (iefode) [inline inference-engine/tests/functional/plugin/cpu/shared_tests_instances/single_layer_tests/scatter_ND_update.cpp:136]: Why commented?

File: inference-engine/tests/functional/plugin/cpu/shared_tests_instances/single_layer_tests/scatter_elements_update.cpp [hunk 21]
```diff
@@ -0,0 +1,110 @@
...
+const std::vector<InferenceEngine::Precision> idxPrecisions = {
+        InferenceEngine::Precision::I32,
+        InferenceEngine::Precision::I64,
+};
+
+const std::vector<std::vector<size_t>> inputShapes = {
+        {8, 11, 11},
+        {12, 10, 19}
+};
+const std::vector<std::vector<size_t>> idxShapes = {
```
Comment (iefode) [inline inference-engine/tests/functional/plugin/cpu/shared_tests_instances/single_layer_tests/scatter_elements_update.cpp:32]: the same comment

File: inference-engine/src/mkldnn_plugin/nodes/mkldnn_scatter_update_node.cpp (hunk 18 above)
Comment (chenhu-wang) [reply inference-engine/src/mkldnn_plugin/nodes/mkldnn_scatter_update_node.cpp:347]: Make muck sense. thank you Dmitry. fixed.

File: ngraph/src/ngraph/op/scatter_elements_update.cpp (hunk 15 above)
[the remark this answers is not in the document]
Comment (lazarevevgeny) [reply ngraph/src/ngraph/op/scatter_elements_update.cpp:252]: @ilyachur was working on cleaning up some reference implementations to reduce the nGraph library binary size. The probability that we will have i8 and I16 as input is very low, so do not extend support for them.

[the remark this answers is not in the document]
Comment (chenhu-wang) [reply ngraph/src/ngraph/op/scatter_elements_update.cpp]: @lazarevevgeny @ilyachur Thank you for clarification. Have moved extension and test for these data type!

File: inference-engine/tests/functional/plugin/cpu/shared_tests_instances/single_layer_tests/scatter_ND_update.cpp (hunk 19 above)
Comment (chenhu-wang) [reply inference-engine/tests/functional/plugin/cpu/shared_tests_instances/single_layer_tests/scatter_ND_update.cpp:84]: Thank you. I will look into.

File: inference-engine/tests/functional/plugin/cpu/shared_tests_instances/single_layer_tests/scatter_ND_update.cpp (hunk 20 above)
Comment (chenhu-wang) [reply inference-engine/tests/functional/plugin/cpu/shared_tests_instances/single_layer_tests/scatter_ND_update.cpp:134]: > Thank you for pointing out. I am implementing ScatterUpdate layers together. When I run single layer test, I showed ScatterNDUpdate and ScatterElementsUpdate ref are not support yet in ngraph. so I [commented out the test](https://github.com/openvinotoolkit/openvino/blob/e891d94ba09adb93bc8dbfadd59f51db8d6047ea/inference-engine/tests/functional/plugin/cpu/shared_tests_instances/single_layer_tests/scatter_ND_update.cpp#L135). and will open when support is available in ngraph.

File: inference-engine/tests/functional/plugin/cpu/shared_tests_instances/single_layer_tests/scatter_update.cpp [hunk 22]
```diff
@@ -0,0 +1,149 @@
...
+        ::testing::ValuesIn(inputShapes5D),
+        ::testing::ValuesIn(idxShapes1D),
+        ::testing::ValuesIn(idxValue1D),
+        ::testing::ValuesIn(updateShapes5D_N),
+        ::testing::ValuesIn(axisValue5D_N),
+        ::testing::ValuesIn(inputPrecisions),
+        ::testing::ValuesIn(idxPrecisions),
+        ::testing::Values(CommonTestUtils::DEVICE_CPU)
+);
+// open after ngraph reference implementation merged
```
[the remark this answers is not in the document]
Comment (lazarevevgeny) [reply inference-engine/tests/functional/plugin/cpu/shared_tests_instances/single_layer_tests/scatter_update.cpp:139]: Nobody is working on them now. But the work on finding the person to implement them is in progress.

File: docs/ops/movement/ScatterUpdate_3.md (hunk 17 above)
Comment (lazarevevgeny) [reply docs/ops/movement/ScatterUpdate_3.md:12]: Yes, this is a correct change!
@chenhu-wang , thank you for noticing the bug!

File: inference-engine/tests/functional/plugin/cpu/shared_tests_instances/single_layer_tests/scatter_ND_update.cpp (hunk 19 above)
Comment (chenhu-wang) [reply inference-engine/tests/functional/plugin/cpu/shared_tests_instances/single_layer_tests/scatter_ND_update.cpp]: @iefode I have combined and simplified, thank you!

File: inference-engine/tests/functional/plugin/cpu/shared_tests_instances/single_layer_tests/scatter_elements_update.cpp (hunk 21 above)
Comment (chenhu-wang) [reply inference-engine/tests/functional/plugin/cpu/shared_tests_instances/single_layer_tests/scatter_elements_update.cpp]: fixed!

File: inference-engine/tests/functional/plugin/cpu/shared_tests_instances/single_layer_tests/scatter_update.cpp (hunk 22 above)
[the remark this answers is not in the document]
Comment (iefode) [reply inference-engine/tests/functional/plugin/cpu/shared_tests_instances/single_layer_tests/scatter_update.cpp:44]: If nobody implements reference in NGraph, we should do it ourselves. I am talking about the JIRA task description:
```
* Add test definition to the shared part of plugin functional tests
* Create instance for CPU plugin
* Create Reference in Ngraph
```

@lazarevevgeny Is it possible to find a person who will implement the reference in Ngraph? If not, @chenhu-wang @dmitry-gorokhov Could you implement the reference?

[the remark this answers is not in the document]
Comment (dmitry-gorokhov) [reply inference-engine/tests/functional/plugin/cpu/shared_tests_instances/single_layer_tests/scatter_update.cpp:44]: @iefode Which one JIRA task? 
The PR covers only following tickets: 29131, 30364, 26814 which assumes only CPU implementation. I am against blocking this PR because of lack of NG reference. 
We will discuss who will be able to implement reference but it will be done separately.

[the remark this answers is not in the document]
Comment (iefode) [reply inference-engine/tests/functional/plugin/cpu/shared_tests_instances/single_layer_tests/scatter_update.cpp:44]: 33266 -> `[IE TESTS] Add to the singe layer tests "ScatterUpdate"`. I am OK to do it separately

[the remark this answers is not in the document]
Comment (lazarevevgeny) [reply inference-engine/tests/functional/plugin/cpu/shared_tests_instances/single_layer_tests/scatter_update.cpp:44]: I agree with @dmitry-gorokhov, let's add nGraph reference implementation in a separate PR. But please wait when @rkazants check the model e2e.

@rkazants , when do you plan to finish the check?

[the remark this answers is not in the document]
Comment (rkazants) [reply inference-engine/tests/functional/plugin/cpu/shared_tests_instances/single_layer_tests/scatter_update.cpp:44]: > I agree with @dmitry-gorokhov, let's add nGraph reference implementation in a separate PR. But please wait when @rkazants check the model e2e.
> 
> @rkazants , when do you plan to finish the check?

today - earlier tomorrow

[the remark this answers is not in the document]
Comment (chenhu-wang) [reply inference-engine/tests/functional/plugin/cpu/shared_tests_instances/single_layer_tests/scatter_update.cpp:44]: @rkazants , just a kind reminder, better to use latest commit to avoid regression check.

[the remark this answers is not in the document]
Comment (rkazants) [reply inference-engine/tests/functional/plugin/cpu/shared_tests_instances/single_layer_tests/scatter_update.cpp:44]: > @rkazants , just a kind reminder, better to use latest commit to avoid regression check.

Good news, e2e test for Wide and Deep model with ScatterNDUpdate passed. Thanks a lot!

File: ngraph/src/ngraph/op/scatter_elements_update.cpp (hunk 15 above)
Comment (chenhu-wang) [reply ngraph/src/ngraph/op/scatter_elements_update.cpp]: @GlebKazantaev,  do you agree with this replay for your request?
https://github.com/openvinotoolkit/openvino/pull/909#discussion_r450228840
https://github.com/openvinotoolkit/openvino/pull/909#discussion_r450251654

Comment (chenhu-wang) [reply ngraph/src/ngraph/op/scatter_elements_update.cpp]: @GlebKazantaev ,
> @ilyachur was working on cleaning up some reference implementations to reduce the nGraph library binary size. The probability that we will have i8 and I16 as input is very low, so do not extend support for them.



> @lazarevevgeny @ilyachur Thank you for clarification. Have moved extension and test for these data type!

**Comment by dmitry-gorokhov** [comment]
@GlebKazantaev @iefode  Do you have any comments? I would like to merge the PR.
